File: qqbot-module-database-support/src/main/java/com/mieai/qqbot/app/database/DatabaseProfile.java
package com.mieai.qqbot.app.database;

import com.mieai.qqbot.admin.database.DatabasePassword;
import com.mieai.qqbot.admin.database.DatabaseSettings;
import com.mieai.qqbot.admin.database.DatabaseSslMode;
import com.mieai.qqbot.admin.database.DatabaseType;
import java.nio.file.Path;
import java.time.Duration;
import java.util.Arrays;
import java.util.Objects;

final class DatabaseProfile implements AutoCloseable {
    private final DatabaseType type;
    private final Path sqlitePath;
    private final Duration busyTimeout;
    private final String host;
    private final int port;
    private final String databaseName;
    private final String username;
    private char[] password;
    private final DatabaseSslMode sslMode;
    private final Duration connectTimeout;

    private DatabaseProfile(
            DatabaseType type,
            Path sqlitePath,
            Duration busyTimeout,
            String host,
            int port,
            String databaseName,
            String username,
            char[] password,
            DatabaseSslMode sslMode,
            Duration connectTimeout) {
        this.type = Objects.requireNonNull(type, "type must not be null");
        this.sqlitePath = sqlitePath;
        this.busyTimeout = busyTimeout;
        this.host = host;
        this.port = port;
        this.databaseName = databaseName;
        this.username = username;
        this.password = password == null ? null : password.clone();
        this.sslMode = sslMode;
        this.connectTimeout = connectTimeout;
        validate();
    }

    static DatabaseProfile sqlite(Path path, Duration busyTimeout) {
        return new DatabaseProfile(
                DatabaseType.SQLITE,
                Objects.requireNonNull(path, "path must not be null").toAbsolutePath().normalize(),
                Objects.requireNonNull(busyTimeout, "busyTimeout must not be null"),
                null,
                0,
                null,
                null,
                null,
                null,
                null);
    }

    static DatabaseProfile server(
            DatabaseType type,
            String host,
            int port,
            String databaseName,
            String username,
            char[] password,
            DatabaseSslMode sslMode,
            Duration connectTimeout) {
        if (type == DatabaseType.SQLITE) {
            throw new IllegalArgumentException("server profile type must not be SQLITE");
        }
        return new DatabaseProfile(
                type,
                null,
                null,
                requireText(host, "host"),
                port,
                requireText(databaseName, "databaseName"),
                requireText(username, "username"),
                Objects.requireNonNull(password, "password must not be null"),
                Objects.requireNonNull(sslMode, "sslMode must not be null"),
                Objects.requireNonNull(connectTimeout, "connectTimeout must not be null"));
    }

    static DatabaseProfile fromSettings(DatabaseSettings settings, char[] resolvedPassword) {
        Objects.requireNonNull(settings, "settings must not be null");
        if (settings.type() == DatabaseType.SQLITE) {
            return sqlite(Path.of(settings.sqlitePath()), Duration.ofMillis(settings.busyTimeoutMs()));
        }
        return server(
                settings.type(),
                settings.host(),
                settings.port(),
                settings.databaseName(),
                settings.username(),
                resolvedPassword,
                settings.sslMode(),
                Duration.ofMillis(settings.connectTimeoutMs()));
    }

    DatabaseType type() {
        return type;
    }

    Path sqlitePath() {
        return sqlitePath;
    }

    Duration busyTimeout() {
        return busyTimeout;
    }

    String host() {
        return host;
    }

    int port() {
        return port;
    }

    String databaseName() {
        return databaseName;
    }

    String username() {
        return username;
    }

    DatabaseSslMode sslMode() {
        return sslMode;
    }

    Duration connectTimeout() {
        return connectTimeout;
    }

    synchronized char[] copyPassword() {
        if (type == DatabaseType.SQLITE) {
            return null;
        }
        if (password == null) {
            throw new IllegalStateException("database password has been destroyed");
        }
        return password.clone();
    }

    boolean hasPassword() {
        return type != DatabaseType.SQLITE;
    }

    boolean sameCredentialScope(DatabaseSettings settings) {
        return settings != null
                && type == settings.type()
                && type != DatabaseType.SQLITE
                && Objects.equals(host, settings.host())
                && port == settings.port()
                && Objects.equals(databaseName, settings.databaseName())
                && Objects.equals(username, settings.username());
    }

    DatabaseProfile copy() {
        char[] copiedPassword = copyPassword();
        try {
            return type == DatabaseType.SQLITE
                    ? sqlite(sqlitePath, busyTimeout)
                    : server(type, host, port, databaseName, username, copiedPassword, sslMode, connectTimeout);
        } finally {
            if (copiedPassword != null) {
                Arrays.fill(copiedPassword, '\0');
            }
        }
    }

    @Override
    public synchronized void close() {
        if (password != null) {
            Arrays.fill(password, '\0');
            password = null;
        }
    }

    @Override
    public String toString() {
        return "DatabaseProfile[type=" + type + ", password=<redacted>]";
    }

    private void validate() {
        if (type == DatabaseType.SQLITE) {
            if (sqlitePath == null || busyTimeout == null || busyTimeout.isNegative() || busyTimeout.isZero()) {
                throw new IllegalArgumentException("SQLite path and positive busy timeout are required");
            }
            return;
        }
        if (port < 1 || port > 65535) {
            throw new IllegalArgumentException("database port must be between 1 and 65535");
        }
        if (connectTimeout.isNegative() || connectTimeout.isZero()) {
            throw new IllegalArgumentException("connect timeout must be positive");
        }
        if (host.codePoints().anyMatch(Character::isWhitespace)
                || host.indexOf('/') >= 0
                || host.indexOf('?') >= 0
                || host.indexOf('#') >= 0) {
            throw new IllegalArgumentException("database host contains unsupported characters");
        }
    }

    private static String requireText(String value, String name) {
        Objects.requireNonNull(value, name + " must not be null");
        String normalized = value.strip();
        if (normalized.isEmpty()) {
            throw new IllegalArgumentException(name + " must not be blank");
        }
        return normalized;
    }
}

File: qqbot-module-database-support/src/main/kotlin/com/mieai/qqbot/app/database/DatabaseProfile.kt
package com.mieai.qqbot.app.database

import com.mieai.qqbot.admin.database.DatabaseSettings
import com.mieai.qqbot.admin.database.DatabaseSslMode
import com.mieai.qqbot.admin.database.DatabaseType
import java.nio.file.Path
import java.time.Duration
import java.util.Arrays

class DatabaseProfile private constructor(
    val type: DatabaseType,
    val sqlitePath: Path?,
    val busyTimeout: Duration?,
    val host: String?,
    val port: Int,
    val databaseName: String?,
    val username: String?,
    password: CharArray?,
    val sslMode: DatabaseSslMode?,
    val connectTimeout: Duration?,
) : AutoCloseable {
    private var password: CharArray? = password?.clone()

    init {
        validate()
    }

    @Synchronized
    fun copyPassword(): CharArray? {
        if (type == DatabaseType.SQLITE) return null
        return password?.clone() ?: throw IllegalStateException("database password has been destroyed")
    }

    fun hasPassword(): Boolean = type != DatabaseType.SQLITE

    fun sameCredentialScope(settings: DatabaseSettings?): Boolean = settings != null &&
        type == settings.type &&
        type != DatabaseType.SQLITE &&
        host == settings.host &&
        port == settings.port &&
        databaseName == settings.databaseName &&
        username == settings.username

    fun copy(): DatabaseProfile {
        val copiedPassword = copyPassword()
        try {
            return if (type == DatabaseType.SQLITE) {
                sqlite(requireNotNull(sqlitePath), requireNotNull(busyTimeout))
            } else {
                server(
                    type,
                    requireNotNull(host),
                    port,
                    requireNotNull(databaseName),
                    requireNotNull(username),
                    requireNotNull(copiedPassword),
                    requireNotNull(sslMode),
                    requireNotNull(connectTimeout),
                )
            }
        } finally {
            copiedPassword?.let { Arrays.fill(it, '\u0000') }
        }
    }

    @Synchronized
    override fun close() {
        password?.let { Arrays.fill(it, '\u0000') }
        password = null
    }

    override fun toString(): String = "DatabaseProfile[type=$type, password=<redacted>]"

    private fun validate() {
        if (type == DatabaseType.SQLITE) {
            require(sqlitePath != null && busyTimeout != null && !busyTimeout.isNegative && !busyTimeout.isZero) {
                "SQLite path and positive busy timeout are required"
            }
            return
        }
        require(port in 1..65535) { "database port must be between 1 and 65535" }
        val timeout = requireNotNull(connectTimeout)
        require(!timeout.isNegative && !timeout.isZero) { "connect timeout must be positive" }
        val serverHost = requireNotNull(host)
        require(
            serverHost.codePoints().noneMatch(Character::isWhitespace) &&
                '/' !in serverHost && '?' !in serverHost && '#' !in serverHost,
        ) { "database host contains unsupported characters" }
    }

    companion object {
        fun sqlite(path: Path, busyTimeout: Duration): DatabaseProfile = DatabaseProfile(
            DatabaseType.SQLITE,
            path.toAbsolutePath().normalize(),
            busyTimeout,
            null,
            0,
            null,
            null,
            null,
            null,
            null,
        )

        fun server(
            type: DatabaseType,
            host: String,
            port: Int,
            databaseName: String,
            username: String,
            password: CharArray,
            sslMode: DatabaseSslMode,
            connectTimeout: Duration,
        ): DatabaseProfile {
            require(type != DatabaseType.SQLITE) { "server profile type must not be SQLITE" }
            return DatabaseProfile(
                type,
                null,
                null,
                requireText(host, "host"),
                port,
                requireText(databaseName, "databaseName"),
                requireText(username, "username"),
                password,
                sslMode,
                connectTimeout,
            )
        }

        fun fromSettings(settings: DatabaseSettings, resolvedPassword: CharArray?): DatabaseProfile =
            if (settings.type == DatabaseType.SQLITE) {
                sqlite(
                    Path.of(requireNotNull(settings.sqlitePath)),
                    Duration.ofMillis(requireNotNull(settings.busyTimeoutMs)),
                )
            } else {
                server(
                    settings.type,
                    requireNotNull(settings.host),
                    requireNotNull(settings.port),
                    requireNotNull(settings.databaseName),
                    requireNotNull(settings.username),
                    requireNotNull(resolvedPassword),
                    requireNotNull(settings.sslMode),
                    Duration.ofMillis(requireNotNull(settings.connectTimeoutMs)),
                )
            }

        private fun requireText(value: String, name: String): String {
            val normalized = value.trim()
            require(normalized.isNotEmpty()) { "$name must not be blank" }
            return normalized
        }
    }
}
